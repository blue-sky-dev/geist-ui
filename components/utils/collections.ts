import React, { ReactNode } from 'react'

export const getId = () => {
  return Math.random().toString(32).slice(2, 10)
}

export const hasChild = (
  children: ReactNode | undefined,
  child: React.ElementType
): boolean => {
  const types = React.Children.map(children, item => {
    if (!React.isValidElement(item)) return null
    return item.type
  })
  
  return (types || []).includes(child)
}

export const pickChild = (
  children: ReactNode | undefined,
  targetChild: React.ElementType
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = []
  const withoutTargetChildren = React.Children.map(children, item => {
    if (!React.isValidElement(item)) return item
    if (item.type === targetChild) {
      target.push(item)
      return null
    }
    return item
  })
  
  const targetChildren = target.length >= 0 ? target : undefined
  
  return [withoutTargetChildren, targetChildren]
}

export const pickChildByProps = (
  children: ReactNode | undefined,
  key: string,
  value: any,
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = []
  const withoutPropChildren = React.Children.map(children, item => {
    if (!React.isValidElement(item)) return null
    if (!item.props) return item
    if (item.props[key] === value) {
      target.push(item)
      return null
    }
    return item
  })
  
  const targetChildren = target.length >= 0 ? target : undefined
  
  return [withoutPropChildren, targetChildren]
}

export const pickChildrenFirst = (
  children: ReactNode | undefined,
): ReactNode | undefined => {
  return React.Children.toArray(children)[0]
}

export const setChildrenProps = (
  children: ReactNode | undefined,
  props: object = {},
  targetComponents: Array<React.ElementType> = [],
): ReactNode | undefined => {
  if (React.Children.count(children) === 0) return []
  const allowAll = targetComponents.length === 0
  const clone = (child: React.ReactElement, props = {}) => React.cloneElement(child, props)

  return React.Children.map(children, item => {
    if (!React.isValidElement(item)) return item
    if (allowAll) return clone(item, props)

    const isAllowed = targetComponents.find(child => child === item.type)
    if (isAllowed) return clone(item, props)
    return item
  })
}

export const setChildrenIndex = (
  children: ReactNode | undefined,
  targetComponents: Array<React.ElementType> = [],
): ReactNode | undefined => {
  if (React.Children.count(children) === 0) return []
  const allowAll = targetComponents.length === 0
  const clone = (child: React.ReactElement, props = {}) => React.cloneElement(child, props)
  let index = 0
  
  return React.Children.map(children, item => {
    if (!React.isValidElement(item)) return item
    index = index + 1
    if (allowAll) return clone(item, { index })
    
    const isAllowed = targetComponents.find(child => child === item.type)
    if (isAllowed) return clone(item, { index })
    index = index - 1
    return item
  })
}

export type ShapeType = {
  width: number
  height: number
}

export const getRealShape = (el: HTMLElement | null): ShapeType => {
  const defaultShape: ShapeType = { width: 0, height: 0 }
  if (!el || typeof window === 'undefined') return defaultShape
  
  const rect = el.getBoundingClientRect()
  const { width, height } = window.getComputedStyle(el)
  
  const getCSSStyleVal = (str: string, parentNum: number) => {
    if (!str) return 0
    const strVal = str.includes('px') ? +str.split('px')[0]
      : str.includes('%') ? +str.split('%')[0] * parentNum * 0.01 : str
    
    return Number.isNaN(+strVal) ? 0 : +strVal
  }

  return {
    width: getCSSStyleVal(`${width}`, rect.width),
    height: getCSSStyleVal(`${height}`, rect.height),
  }
}
