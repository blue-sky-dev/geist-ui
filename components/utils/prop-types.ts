
export const tuple = <T extends string[]>(...args: T) => args

const buttonTypes = tuple(
  'default',
  'secondary',
  'success',
  'warning',
  'error',
  'abort',
)

const normalSizes = tuple(
  'mini',
  'small',
  'medium',
  'large',
)

const normalTypes = tuple(
  'default',
  'secondary',
  'success',
  'warning',
  'error',
)

const themeTypes = tuple(
  'dark',
  'light',
)

const snippetTypes = tuple(
  'default',
  'secondary',
  'success',
  'warning',
  'error',
  'dark',
  'lite',
)

const copyTypes = tuple(
  'default',
  'slient',
  'prevent',
)

export type ButtonTypes = typeof buttonTypes[number]

export type NormalSizes = typeof normalSizes[number]

export type NormalTypes = typeof normalTypes[number]

export type ThemeTypes = typeof themeTypes[number]

export type SnippetTypes = typeof snippetTypes[number]

export type CopyTypes = typeof copyTypes[number]
