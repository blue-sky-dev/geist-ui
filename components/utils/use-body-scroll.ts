import useBodyScroll from '../utils-shared-deprecated/use-body-scroll'

export default useBodyScroll
