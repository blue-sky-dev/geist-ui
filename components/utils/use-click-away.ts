import { default as useClickAway } from '../utils-shared/use-click-away'

export default useClickAway
