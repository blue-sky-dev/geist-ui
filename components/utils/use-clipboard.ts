import useClipboard from '../utils-shared/use-clipboard'

export default useClipboard
