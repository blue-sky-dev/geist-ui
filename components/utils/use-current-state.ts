import useCurrentState from '../utils-shared/use-current-state'

export default useCurrentState
