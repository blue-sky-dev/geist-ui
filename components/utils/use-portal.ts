import { useEffect, useState } from 'react'
import useSSR from '../utils/use-ssr'

const getId = () => {
  return Math.random().toString(32).slice(2, 10)
}

const createElement = (id: string): HTMLElement => {
  const el = document.createElement('div')
  el.setAttribute('id', id)
  return el
}

const usePortal = (selectId: string = getId()): Element | null => {
  const id = `zeit-ui-${selectId}`
  const { isBrowser } = useSSR()
  const [elSnapshot, setElSnapshot] = useState<Element | null>(isBrowser ? createElement(id) : null)
  
  useEffect(() => {
    const hasElement = document.querySelector(`#${id}`)
    const el = hasElement || createElement(id)

    if (!hasElement) {
      document.body.appendChild(el)
    }
    setElSnapshot(el)
  
    return () => {
      const node = document.getElementById(id)
      if (node) {
        // document.body.removeChild(node)
      }
    }
  }, [])

  return elSnapshot
}

export default usePortal
