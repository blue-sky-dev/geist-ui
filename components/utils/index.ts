import { default as useBodyScroll } from './use-body-scroll'
import { default as useClickAnywhere } from './use-click-anywhere'
import { default as useClickAway } from './use-click-away'
import { default as useClipboard } from './use-clipboard'
import { default as useCurrentState } from './use-current-state'
import { default as useRealShape } from './use-real-shape'
import { default as useResize } from './use-resize'
import { default as useSSR } from './use-ssr'
import { default as withDefaults } from './with-defaults'

export default {
  withDefaults,
  useBodyScroll,
  useClickAnywhere,
  useClickAway,
  useClipboard,
  useCurrentState,
  useRealShape,
  useResize,
  useSSR,
}
