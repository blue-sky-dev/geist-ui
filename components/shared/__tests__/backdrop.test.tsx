import React from 'react'
import { mount } from 'enzyme'
import Backdrop from '../backdrop'
import { nativeEvent, updateWrapper } from 'tests/utils'

describe('Backdrop', () => {
  it('should render correctly', () => {
    const wrapper = mount(
      <Backdrop visible>
        <span>test-value</span>
      </Backdrop>,
    )
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should be shown or hidden by prop', async () => {
    const wrapper = mount(
      <Backdrop>
        <span>test-value</span>
      </Backdrop>,
    )
    expect(wrapper.find('.backdrop').length).toBe(0)
    wrapper.setProps({ visible: true })
    await updateWrapper(wrapper, 350)
    expect(wrapper.find('.backdrop').length).not.toBe(0)
  })

  it('background click events should be captured', () => {
    const handler = jest.fn()
    const wrapper = mount(
      <Backdrop onClick={handler} visible>
        <span>test-value</span>
      </Backdrop>,
    )
    wrapper.find('.backdrop').simulate('click', nativeEvent)
    expect(handler).toHaveBeenCalled()
    handler.mockRestore()
  })

  it('should be no error when handler missing', () => {
    const wrapper = mount(
      <Backdrop visible>
        <span>test-value</span>
      </Backdrop>,
    )
    wrapper.find('.backdrop').simulate('click', nativeEvent)
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should be prevent event from the container', () => {
    const handler = jest.fn()
    const wrapper = mount(
      <Backdrop onClick={handler} visible>
        <span>test-value</span>
      </Backdrop>,
    )
    wrapper.find('.content').simulate('click', nativeEvent)
    wrapper.find('.offset').simulate('click', nativeEvent)
    expect(handler).not.toHaveBeenCalled()
    handler.mockRestore()
  })

  it('content should be offset', () => {
    const wrapper = mount(
      <Backdrop visible>
        <span>test-value</span>
      </Backdrop>,
    )
    const notOffset = wrapper.html()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({ offsetY: '100' })
    expect(wrapper.html()).toMatchSnapshot()
    expect(notOffset).not.toEqual(wrapper.html())
  })
})
