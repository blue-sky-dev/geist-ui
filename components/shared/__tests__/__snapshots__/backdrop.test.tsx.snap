// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop content should be offset 1`] = `
"<div class=\\"backdrop  transition-enter\\"><div class=\\"layer\\"></div><div class=\\"content\\"><span>test-value</span></div><div class=\\"offset\\"></div><style>
            .backdrop {
              position: fixed;
              top: 0;
              left: 0;
              display: flex;
              align-content: center;
              align-items: center;
              flex-direction: column;
              justify-content: center;
              height: 100vh;
              width: 100vw;
              overflow: auto;
              z-index: 1000;
            }

            .content {
              display: flex;
              align-items: center;
              justify-content: center;
              position: relative;
              z-index: 1001;
              outline: none;
            }

            .offset {
              height: 0vh;
              opacity: 0;
              background-color: transparent;
            }

            .layer {
              position: fixed;
              top: 0;
              left: 0;
              right: 0;
              bottom: 0;
              width: 100%;
              height: 100%;
              opacity: 0.25;
              background-color: black;
              transition: opacity 0.35s cubic-bezier(0.4, 0, 0.2, 1);
              pointer-events: none;
              z-index: 1000;
            }
          </style></div>"
`;

exports[`Backdrop content should be offset 2`] = `
"<div class=\\"backdrop  transition-enter\\"><div class=\\"layer\\"></div><div class=\\"content\\"><span>test-value</span></div><div class=\\"offset\\"></div><style>
            .backdrop {
              position: fixed;
              top: 0;
              left: 0;
              display: flex;
              align-content: center;
              align-items: center;
              flex-direction: column;
              justify-content: center;
              height: 100vh;
              width: 100vw;
              overflow: auto;
              z-index: 1000;
            }

            .content {
              display: flex;
              align-items: center;
              justify-content: center;
              position: relative;
              z-index: 1001;
              outline: none;
            }

            .offset {
              height: 100vh;
              opacity: 0;
              background-color: transparent;
            }

            .layer {
              position: fixed;
              top: 0;
              left: 0;
              right: 0;
              bottom: 0;
              width: 100%;
              height: 100%;
              opacity: 0.25;
              background-color: black;
              transition: opacity 0.35s cubic-bezier(0.4, 0, 0.2, 1);
              pointer-events: none;
              z-index: 1000;
            }
          </style></div>"
`;

exports[`Backdrop should render correctly 1`] = `
"<div class=\\"backdrop  transition-enter\\"><div class=\\"layer\\"></div><div class=\\"content\\"><span>test-value</span></div><div class=\\"offset\\"></div><style>
            .backdrop {
              position: fixed;
              top: 0;
              left: 0;
              display: flex;
              align-content: center;
              align-items: center;
              flex-direction: column;
              justify-content: center;
              height: 100vh;
              width: 100vw;
              overflow: auto;
              z-index: 1000;
            }

            .content {
              display: flex;
              align-items: center;
              justify-content: center;
              position: relative;
              z-index: 1001;
              outline: none;
            }

            .offset {
              height: 0vh;
              opacity: 0;
              background-color: transparent;
            }

            .layer {
              position: fixed;
              top: 0;
              left: 0;
              right: 0;
              bottom: 0;
              width: 100%;
              height: 100%;
              opacity: 0.25;
              background-color: black;
              transition: opacity 0.35s cubic-bezier(0.4, 0, 0.2, 1);
              pointer-events: none;
              z-index: 1000;
            }
          </style></div>"
`;
